### PR TITLE
Add a few functionalities to minecart extension and fixed collision h…

### PR DESCRIPTION
**Pull request by liach**
…andler unsettable

Signed-off-by: liach <liach@users.noreply.github.com>

### CODE REVIEW

**Comment by tterrag1098** [comment]
These patches need justifications. What are you doing that needs this?

File: patches/minecraft/net/minecraft/entity/item/EntityMinecart.java.patch [hunk 1]
```diff
@@ -9,15 +9,7 @@
     private static final DataParameter<Integer> field_184265_a = EntityDataManager.func_187226_a(EntityMinecart.class, DataSerializers.field_187192_b);
     private static final DataParameter<Integer> field_184266_b = EntityDataManager.func_187226_a(EntityMinecart.class, DataSerializers.field_187192_b);
     private static final DataParameter<Float> field_184267_c = EntityDataManager.func_187226_a(EntityMinecart.class, DataSerializers.field_187193_c);
-@@ -59,6 +59,7 @@
-    private double field_70507_ar;
-    @OnlyIn(Dist.CLIENT)
-    private double field_70506_as;
-+   private boolean canBePushed = true;
```
Comment (liach) [inline patches/minecraft/net/minecraft/entity/item/EntityMinecart.java.patch:16]: Moved to forge addition chunk below

File: patches/minecraft/net/minecraft/entity/item/EntityMinecart.java.patch [hunk 2]
```diff
@@ -31,7 +23,25 @@
     }
  
     public double func_70042_X() {
-@@ -260,7 +262,7 @@
+@@ -123,7 +124,7 @@
+    }
+ 
+    public boolean func_70097_a(DamageSource p_70097_1_, float p_70097_2_) {
+-      if (!this.field_70170_p.field_72995_K && !this.field_70128_L) {
++      if (!this.field_70170_p.field_72995_K && this.func_70089_S()) {
```
Comment (liach) [inline patches/minecraft/net/minecraft/entity/item/EntityMinecart.java.patch:31]: Removed deprecated `removed` field usage

File: patches/minecraft/net/minecraft/entity/item/EntityMinecart.java.patch [hunk 3]
```diff
@@ -31,7 +23,25 @@
...
+          if (this.func_180431_b(p_70097_1_)) {
+             return false;
+          } else {
+@@ -169,7 +170,7 @@
+    }
+ 
+    public boolean func_70067_L() {
+-      return !this.field_70128_L;
++      return this.func_70089_S();
```
Comment (liach) [inline patches/minecraft/net/minecraft/entity/item/EntityMinecart.java.patch:40]: Same deprecated field `Entity#removed` usage

File: patches/minecraft/net/minecraft/entity/item/EntityMinecart.java.patch [hunk 4]
```diff
@@ -90,9 +100,12 @@
...
        if (this.field_70122_E) {
-          this.field_70159_w *= 0.5D;
-          this.field_70181_x *= 0.5D;
-          this.field_70179_y *= 0.5D;
+-         this.field_70159_w *= 0.5D;
+-         this.field_70181_x *= 0.5D;
+-         this.field_70179_y *= 0.5D;
++         this.field_70159_w *= getDragOnGround();
++         this.field_70181_x *= getDragOnGround();
++         this.field_70179_y *= getDragOnGround();
```
Comment (liach) [inline patches/minecraft/net/minecraft/entity/item/EntityMinecart.java.patch:108]: Addition: allow modded minecarts to specify ground drag in addition to air drag

File: patches/minecraft/net/minecraft/entity/item/EntityMinecart.java.patch [hunk 5]
```diff
@@ -223,7 +236,15 @@
                          p_70108_1_.field_70159_w *= (double)0.2F;
                          p_70108_1_.field_70179_y *= (double)0.2F;
                          p_70108_1_.func_70024_g(this.field_70159_w + d0, 0.0D, this.field_70179_y + d1);
-@@ -782,6 +796,32 @@
+@@ -746,6 +759,7 @@
+       return this.field_70180_af.func_187225_a(field_184266_b);
+    }
+ 
++   @Deprecated // Forge: Use individual minecart methods instead
+    public abstract EntityMinecart.Type func_184264_v();
```
Comment (liach) [inline patches/minecraft/net/minecraft/entity/item/EntityMinecart.java.patch:244]: Enhances mod compatibility as we discourage using `Type` enum for minecarts but use methods instead

File: src/main/java/net/minecraftforge/common/extensions/ExtensionFields.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.common.extensions;
+
+import net.minecraftforge.common.IMinecartCollisionHandler;
+
+final class ExtensionFields
+{
+    static IMinecartCollisionHandler minecartCollisionHandler = null;
```
Comment (liach) [inline src/main/java/net/minecraftforge/common/extensions/ExtensionFields.java:26]: This field is mutable, unlike constants in interfaces (implied `public static final`).

File: src/main/java/net/minecraftforge/common/extensions/IForgeEntityMinecart.java [hunk 7]
```diff
@@ -37,22 +39,26 @@
...
+    }
+
     /**
      * Gets the current global Minecart Collision handler if none
      * is registered, returns null
      * @return The collision handler or null
      */
-    default IMinecartCollisionHandler getCollisionHandler() {
-        return COLLISIONS;
+    default  @Nullable IMinecartCollisionHandler getCollisionHandler() {
```
Comment (liach) [inline src/main/java/net/minecraftforge/common/extensions/IForgeEntityMinecart.java:60]: Not `static` as to save patch size, or I need to call
`net.minecraftforge.common.extensions.IForgeEntityMinecart.getCollisionHandler()` instead.

**Review by liach — COMMENTED** [review]
@tterrag1098 These are my explanation for the patches.

**Comment by CovertJaguar** [comment]
I'm going to put the kabosh on this as I'm currently, at the request of Lex, doing a complete review and use case analysis of the minecart/track patches.

I'm going to a place with no electricity for the next couple weeks however, and while I'm about half way through with the review, it probably won't be finished before mid-June.

My current findings indicate that we could likely reduce the number of patches to EntityMinecart and BlockRailBase by nearly three quarters of the current total.

Current Draft of Analysis:
https://pastebin.com/ZTFS3Bh2

I will eventually be writing this up into a proper issue/PR when I have the time in a couple weeks.

File: patches/minecraft/net/minecraft/entity/item/EntityMinecart.java.patch (hunk 5 above)
Comment (mezz) [reply patches/minecraft/net/minecraft/entity/item/EntityMinecart.java.patch:244]: In the comment, please give examples for the methods that should be used instead.

**Review by mezz — CHANGES_REQUESTED** [review]

**Comment by liach** [comment]
Closing as fg3 is glitching out for me again and 1.13 is dead. I hope @CovertJaguar's talk with lex better resolves this issue.

**Pull request closed without merging**